### PR TITLE
[import] Handle importing resource properties of type Union

### LINKED ISSUE #10934: Embedded `PANIC` in generated Python code for imported Azure Frontdoor resource
### What happened?

The Python code generated by Pulumi when importing an Azure Frontdoor resource works for most parts, but the nested `route_configuration` property doesn't contain a proper Python object:

```
route_configuration={
    "backendPool": azure_native.network.SubResourceArgs(
        %!v(PANIC=Format method: interface conversion: model.Expression is *model.TemplateExpression, not *model.LiteralValueExpression),
    "forwardingProtocol": "HttpsOnly",
    "odataType": "#Microsoft.Azure.FrontDoor.Models.FrontdoorForwardingConfiguration",
},
```

Also note the embedded string version of a `PANIC`.

### Steps to reproduce

The click-ops way, create an Azure Classic Frontdoor resource (note: the word `Classic` refers to the Azure naming of the product, not pointing to our Azure Classic provider).

Now try to import it using the ARN:

```
$ pulumi import azure-native:network:FrontDoor test-cdn <arn>
```

### Expected Behavior

Correctly generated Python code for a `azure_native.network.Frontdoor` resource, with this as the correct code for the `route_configuration` section:

```
route_configuration=azure_native.network.ForwardingConfigurationArgs(
    backend_pool=azure_native.network.SubResourceArgs(id="<redacted_arn>"),
    forwarding_protocol="HttpsOnly",
    odata_type="#Microsoft.Azure.FrontDoor.Models.FrontdoorForwardingConfiguration"
)
```

### Actual Behavior

This snippet was generated for this resource:

```
route_configuration={
    "backendPool": azure_native.network.SubResourceArgs(
        %!v(PANIC=Format method: interface conversion: model.Expression is *model.TemplateExpression, not *model.LiteralValueExpression),
    "forwardingProtocol": "HttpsOnly",
    "odataType": "#Microsoft.Azure.FrontDoor.Models.FrontdoorForwardingConfiguration",
},
```

### Output of `pulumi about`

_No response_

### Additional context

Filing this here rather than in pulumi/pulumi-azure-native repo given it involves code generation.

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

### PR DESCRIPTION
**Pull request by Zaid-Ajaj**
# Description

Fixes #10934

## Checklist

- [ ] I have added tests that prove my fix is effective or that my feature works

- [x] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version

### CODE REVIEW

**Comment by Frassle** [comment]
Adding Ian and Justin to review as well, I don't fully get all the PCL codebase yet.

File: pkg/codegen/importer/hcl2.go [hunk 1]
```diff
@@ -425,7 +426,68 @@ func generateValue(typ schema.Type, value resource.PropertyValue) (model.Express
 					})
 				}
 			}
-		} else {
+		case *schema.UnionType:
+			discriminatorPropertyName := objectType.Discriminator
+			selectElementType := func() (*schema.ObjectType, bool) {
+				discriminatorValue, ok := obj[resource.PropertyKey(discriminatorPropertyName)]
+				if !ok {
+					// discriminator property is not present, so we can't select a type
```
Comment (Frassle) [inline pkg/codegen/importer/hcl2.go:434]: I _think_ we need to fallback to the untyped map cases if we fail to work out the union case type correctly? @iwahbe or @justinvp might be able to confirm that, I'm not totally sure what the constraints and expectations of this method are.

File: pkg/codegen/importer/hcl2.go [hunk 2]
```diff
@@ -425,7 +426,68 @@ func generateValue(typ schema.Type, value resource.PropertyValue) (model.Express
...
+						items = append(items, model.ObjectConsItem{
+							Key: &model.LiteralValueExpression{
+								Value: cty.StringVal(p.Name),
+							},
+							Value: x,
+						})
+					}
+				}
+			}
+
```
Comment (iwahbe) [inline pkg/codegen/importer/hcl2.go:489]: I think we need to do something for `!ok`. We can fall back to what we were doing before.

File: pkg/codegen/importer/hcl2.go [hunk 3]
```diff
@@ -425,7 +426,68 @@ func generateValue(typ schema.Type, value resource.PropertyValue) (model.Express
 					})
 				}
 			}
-		} else {
+		case *schema.UnionType:
+			discriminatorPropertyName := objectType.Discriminator
+			selectElementType := func() (*schema.ObjectType, bool) {
```
Comment (iwahbe) [inline pkg/codegen/importer/hcl2.go:431]: Why do we create a lambda here, only to invoke it on the next line? We should either separate out `selectElementType` into its own function or inline it.

File: pkg/codegen/importer/hcl2.go [hunk 4]
```diff
@@ -425,7 +426,68 @@ func generateValue(typ schema.Type, value resource.PropertyValue) (model.Express
...
+
+				discriminatorValueText := discriminatorValue.StringValue()
+				correspondingTypeToken, ok := objectType.Mapping[discriminatorValueText]
+				if ok {
+					for _, elementType := range objectType.ElementTypes {
+						// found the type token
+						// match it against the element type which should be an object
+						elementTypeObject, ok := codegen.UnwrapType(elementType).(*schema.ObjectType)
+						if ok {
+							elementTypeParts := strings.Split(elementTypeObject.Token, ":")
```
Comment (iwahbe) [inline pkg/codegen/importer/hcl2.go:452]: Can we replace this by casting to [`tokens.Type`](https://github.com/pulumi/pulumi/blob/adef574983fa6f4e1ab48e50ef5d1388da1efcdb/sdk/go/common/tokens/tokens.go#L212) and then calling the appropriate methods?

**Review by iwahbe — COMMENTED** [review]
I think it would be simpler to handle this in a more general case. After we set `typ = codegen.UnwrapType(type)`, we could put 
```go
if union, ok := typ.(*schema.UnionType); ok {
  typ = reduceUnion(union, value)
}
```

Reduce union would look like
```
// reduceUnion returns the element type in union that bests fits value.
// If more than one type fits, reduceUnion will return the first type in union.Elements.
// If no type fits, union.DefaultType is returned. The default type may be `nil`.
func reduceUnion(union *schema.UnionType, value resource.PropertyValue) schema.Type
```

`UnionType.Discriminator` is optional, so we will need to reduce based on the shape of the value, not just on the discriminator.

**Comment by Zaid-Ajaj** [comment]
> UnionType.Discriminator is optional, so we will need to reduce based on the shape of the value, not just on the discriminator.

@iwahbe you mean here for each property in the object value, go through each union element type and do a best-effect match to see one object fits? It doesn't seem deterministic imo, maybe in this case we fall back to the `default` case handling. What do you think?

File: pkg/codegen/importer/hcl2.go (hunk 3 above)
Comment (Zaid-Ajaj) [reply pkg/codegen/importer/hcl2.go:431]: Just a stylistic choice to make to more readable. I'll rewrite it.

**Comment by iwahbe** [comment]
> > UnionType.Discriminator is optional, so we will need to reduce based on the shape of the value, not just on the discriminator.
> 
> @iwahbe you mean here for each property in the object value, go through each union element type and do a best-effect match to see one object fits? It doesn't seem deterministic imo, maybe in this case we fall back to the `default` case handling. What do you think?

I mean for each Element type in the union, check if `value` fits. It can't be for for each property in the object value, because lots of unions have element types `[string, SomeObject]` for example. It will be deterministic, but might not always be able to fully narrow the type, here well make a best guess (which must be deterministic). That's ok since `typ` is advisory.

I think we should fall back to the `default` option iff no type fits. I'll amend my comments on the `reduceUnion` function to fit.

**Comment by Zaid-Ajaj** [comment]
@iwahbe I've addressed your comments, can you have another look? 🙏

File: pkg/codegen/importer/hcl2.go [hunk 5]
```diff
@@ -365,11 +366,109 @@ func generatePropertyValue(property *schema.Property, value resource.PropertyVal
...
 
+// reduceUnionType reduces the given union type to a simpler type that potentially matches the value.
+// When the value type is primitive, choose the first element type of the union elements that is of the same type.
+// When the value is an object, use the discriminator to choose the element type.
+func reduceUnionType(schemaUnion *schema.UnionType, value resource.PropertyValue) schema.Type {
+	switch {
+	case value.IsObject():
+		// If the value is an object, use the discriminator to choose the element type.
+		if schemaUnion.Discriminator == "" {
+			// If there is no discriminator, return first type which is an object
```
Comment (iwahbe) [inline pkg/codegen/importer/hcl2.go:377]: I don't agree with this. If there is no discriminator, we need to return the first type it is legal to assign `value` to. This is a structural type check.

For example, given the value `{ foo: 1 }` and `Union[{ foo: string }, { bar: Union[none, string], foo: number }` we should resolve to the second element, not the first.

An example of where this would be necessary is 
https://github.com/pulumi/pulumi-aws-native/blob/edc3d04caddc8996061d1ce2a8b80fbe5236b313/provider/cmd/pulumi-resource-aws-native/schema.json#L13430-L13451.

**Review by iwahbe — COMMENTED** [review]
I left a comment on union resolution. If you want, we can return `nil` for non-discriminated unions and open a new issue to add the correct checking. 

We will definitely want a test here.

File: pkg/codegen/importer/hcl2.go (hunk 5 above)
Comment (Zaid-Ajaj) [reply pkg/codegen/importer/hcl2.go]: I've implemented the structural type check to find the best fit + tests ✅

File: pkg/codegen/importer/hcl2.go [hunk 6]
```diff
@@ -365,11 +366,222 @@ func generatePropertyValue(property *schema.Property, value resource.PropertyVal
...
+
+	case value.IsNumber():
+		// check to see if one of the element types in the union
+		// actually allows a number type and return it
+		for _, t := range schemaUnion.ElementTypes {
+			if t == schema.NumberType {
+				return t
+			}
+		}
+	}
```
Comment (iwahbe) [inline pkg/codegen/importer/hcl2.go:570]: Unions come in two types: discriminated objects and structural. We can simplify this (and handle the missing `case value.IsArray()`) by invoking `valueStructurallyTypedAs`:
```suggestion
	default:
		for _, t := range schemaUnion.ElementTypes {
			if valueStructurallyTypedAs(value, t) {
				return t
			}
		}
	}
```

File: pkg/codegen/importer/hcl2.go [hunk 7]
```diff
@@ -365,11 +366,222 @@ func generatePropertyValue(property *schema.Property, value resource.PropertyVal
 	return generateValue(property.Type, value)
 }
 
+// valueStructurallyTypedAs returns true if the given value is structurally typed as the given schema type.
+func valueStructurallyTypedAs(value resource.PropertyValue, schemaType schema.Type) bool {
+	switch {
```
Comment (iwahbe) [inline pkg/codegen/importer/hcl2.go:371]: We can hoist the union check here, since we need to repeat it for every value anyway:
```suggestion
	if union, ok := schemaType.(*schema.Union); ok {
		schemaType = reduceUnion(union, value)
	}
	switch
```
The reduction is mutually recursive since we want to handle nested discriminated unions.

This lets us remove the `case *schema.UnionType` checks in each branch of the `case value.Is*()`.

File: pkg/codegen/importer/hcl2.go [hunk 8]
```diff
@@ -365,11 +366,222 @@ func generatePropertyValue(property *schema.Property, value resource.PropertyVal
...
+	case value.IsObject():
+		switch arg := schemaType.(type) {
+		case *schema.ObjectType:
+			schemaProperties := make(map[string]schema.Type)
+			for _, schemaProperty := range arg.Properties {
```
Comment (iwahbe) [inline pkg/codegen/importer/hcl2.go:376]: This map is built into `schema.ObjectType`. The API is [here](https://github.com/pulumi/pulumi/blob/16260035bb7c32e5883bbe14061b2a19cf6a45c3/pkg/codegen/schema/schema.go#L256).

File: pkg/codegen/importer/hcl2.go [hunk 9]
```diff
@@ -365,11 +366,222 @@ func generatePropertyValue(property *schema.Property, value resource.PropertyVal
...
+				schemaProperties[schemaProperty.Name] = schemaProperty.Type
```
Comment (iwahbe) [inline pkg/codegen/importer/hcl2.go:377]:
```suggestion
			// If `value.ObjectValue()` is missing a property
			if _, ok := value.ObjectValue()[schemaProperty.Name]; !ok 
				// And that property is not optional
				if _, ok := schemaProperty.Type.(*schema.OptionalType); !ok {
					// We are missing a required property, so the type does not match
					return false
				}
			}
```

File: pkg/codegen/importer/hcl2.go [hunk 10]
```diff
@@ -365,11 +366,222 @@ func generatePropertyValue(property *schema.Property, value resource.PropertyVal
 	return generateValue(property.Type, value)
 }
 
+// valueStructurallyTypedAs returns true if the given value is structurally typed as the given schema type.
+func valueStructurallyTypedAs(value resource.PropertyValue, schemaType schema.Type) bool {
+	switch {
+	case value.IsObject():
+		switch arg := schemaType.(type) {
+		case *schema.ObjectType:
+			schemaProperties := make(map[string]schema.Type)
```
Comment (iwahbe) [inline pkg/codegen/importer/hcl2.go:375]: This map is built into `schema.ObjectType`. The API is [here](https://github.com/pulumi/pulumi/blob/16260035bb7c32e5883bbe14061b2a19cf6a45c3/pkg/codegen/schema/schema.go#L256).

**Review by iwahbe — APPROVED** [review]
I left a last set of comments, but I'm out of corner cases. I appreciate the tests and working with me to add what is essentially a structural type checker to `package importer`.

LGTM!

**Comment by Zaid-Ajaj** [comment]
bors merge

**Comment by AaronFriel** [comment]
bors cancel

**Comment by AaronFriel** [comment]
bors retry

**Comment by AaronFriel** [comment]
bors merge

**Comment by AaronFriel** [comment]
I canceled & requeued as one of the build jobs hanged: https://github.com/pulumi/pulumi/actions/runs/3246406652/jobs/5325125965

**Comment by AaronFriel** [comment]
bors cancel

**Comment by AaronFriel** [comment]
bors merge

**Comment by AaronFriel** [comment]
I would have liked to get this in, but it failed 3 times in CI.

https://github.com/pulumi/pulumi/actions/runs/3246806849/jobs/5326028122

**Comment by AaronFriel** [comment]
bors cancel

**Comment by Zaid-Ajaj** [comment]
@AaronFriel can't seem to decipher why CI is failing or why it would fail in regard to this change, will just retry before digging further into it

**Comment by Zaid-Ajaj** [comment]
bors merge

**Comment by Zaid-Ajaj** [comment]
bors retry
